### PR TITLE
Duplicates handling for IConfig = big refactor, fixes #360

### LINKED ISSUE #360: [Suggestion] Duplicates handling for IConfig
Current implementations of `CompositeXxx` wrappers (`CompositeLogger`, `CompositeDiagnoser` etc) do not have common policy for duplicate values.

For example, the `CompositeDiagnoser` [removes duplicates](https://github.com/dotnet/BenchmarkDotNet/blob/f1f2317dafc663a2fa854ea566ede8befc68ea6d/src/BenchmarkDotNet.Core/Diagnosers/CompositeDiagnoser.cs#L17) but the `CompositeAnalyser` [stores them as-is](https://github.com/dotnet/BenchmarkDotNet/blob/2f85847cdf077e695f0e65b71ed9f1b5faac7d5c/src/BenchmarkDotNet.Core/Analysers/CompositeAnalyser.cs#L14)

**Option one**
Add `.Distinct()` calls at `ConfigExtensions.GetCompositeXxx()` [methods](https://github.com/dotnet/BenchmarkDotNet/blob/24e944ae06101a8051955fcfb0ae6c4413950fdc/src/BenchmarkDotNet.Core/Configs/ConfigExtensions.cs#L16-L20).

**Option two**
Introduce explicit config preparation step that will create full config and remove duplicates at the end. Make config preparation to be responsibility of the `BenchmarkRunner`. 
The `BenchmarkRunnerCore` should receive a not-null prepared config parameter and should not call 
`BenchmarkConverter.GetFullConfig()` by itself. For null configs ArgumentNullException should be thrown.

I vote for option two with two hands as it'll solve console dependency issue at #356 (will allow to remove the `FallbackLogger`). Moreover it will simplify code for my own perftest project as it uses its own [configuration subsystem](https://github.com/rsdn/CodeJam/blob/master/PerfTests/docs/ConfigurationSystem.md) that conflicts with `IConfigSource` attributes.

### PR DESCRIPTION
**Pull request by adamsitnik**
This is a big refactor to solve #360 requested by @ig-sinicyn a long time ago.

My motivation was a few minor bugs and logic for building the final config spread amongst the entire library.

The idea:

1. Make `ManualConfig` dummy. Just a bag of settings with no logic
2. Make all `CompositeXYZ` dummy. They just get something and use it
3. Put all of the logic for building a FinalConfig to one place.
4. Add tests
5. Make the `FinalConfig` immutable
6. If you want to create `CompositeXYZ` only `FinalConfig` can do that for you

### CODE REVIEW

File: src/BenchmarkDotNet/Configs/FinalConfig.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+using BenchmarkDotNet.Order;
+using BenchmarkDotNet.Reports;
+using BenchmarkDotNet.Running;
+using BenchmarkDotNet.Validators;
+using JetBrains.Annotations;
+using RunMode = BenchmarkDotNet.Diagnosers.RunMode;
+
+namespace BenchmarkDotNet.Configs
+{
+    public class FinalConfig : IConfig
```
Comment (AndreyAkinshin) [inline src/BenchmarkDotNet/Configs/FinalConfig.cs]: I don't like the class name. Maybe it should be `ImmutableConfig`?

**Comment by AndreyAkinshin** [comment]
@adamsitnik, in general, the PR looks great. My only concern is about the name of `FinalConfig`. Maybe it should be `ImmutableConfig` or `SealedConfig`?

**Comment by adamsitnik** [comment]
@AndreyAkinshin heh, I also was not sure about the name ;)

Internally we have few configs (global/from args/from attributes). Before we run the benchmarks, we merge all of these configs into one. And this is the config that we run the benchmarks against.

This is why I have called it `Final`. Initially, I have called it `FinalImmutableConfig` but it was too long.

I am not myself a big fan of the `FinalConfig` name, but what I want to express is that this is the config that we run the benchmarks against, no more merging or adding new things. Do you think that `ImmutableConfig` is a better name for it? I am totally open to renaming it or other suggestions.

@ig-sinicyn maybe you have some ideas?

**Comment by AndreyAkinshin** [comment]
I think, `ImmutableConfig` is better because it describes the main "feature" of this config. The class name shouldn't express how we are going to use it in a particular use case, we have field/property names for that.

**Comment by adamsitnik** [comment]
@AndreyAkinshin I have solved the merge conflicts and finished the refactor. I think it's ready for a review

This PR fixes #938, #360 and #463

**Pull request merged**